File: golio/infrastructure/repository/dto/media.go
package dto

import (
	"time"

	"github.com/sunjin110/folio/golio/domain/model"
)

type Media []*Medium

func (m Media) ToSummariesModel(thumbanilURLMap map[string]string) []*model.MediumSummary {
	summaries := make([]*model.MediumSummary, 0, len(m))
	for _, medium := range m {
		summaries = append(summaries, medium.ToSummaryModel(thumbanilURLMap[medium.ID]))
	}
	return summaries
}

type Medium struct {
	ID        string    `db:"id"`
	Path      string    `db:"path"`
	FileType  string    `db:"file_type"`
	CreatedAt time.Time `db:"created_at"`
	UpdatedAt time.Time `db:"updated_at"`
}

func (m *Medium) ToModel(thumbnailURL string, downloadURL string) *model.Medium {
	if m == nil {
		return nil
	}

	return &model.Medium{
		ID:           m.ID,
		FileType:     m.FileType,
		ThumbnailURL: thumbnailURL,
		DownloadURL:  downloadURL,
		CreatedAt:    m.CreatedAt,
		UpdatedAt:    m.UpdatedAt,
	}
}

func (m *Medium) ToSummaryModel(thumbnailURL string) *model.MediumSummary {
	return &model.MediumSummary{
		ID:           m.ID,
		FileType:     m.FileType,
		ThumbnailURL: thumbnailURL,
		CreatedAt:    m.CreatedAt,
		UpdatedAt:    m.UpdatedAt,
	}
}

File: reolio/src/infrastructure/repository/media.ts
import { Medium, MediumSummary } from "@/domain/model/media";
import {
  FindMediumSummariesOutput,
  MediaRepository,
} from "@/domain/repository/media";
import { AuthError, InternalError } from "@/error/error";
import {
  GolioApi,
  MediaGet200Response,
  MediaMediumIdGet200Response,
  MediaPost200Response,
  ResponseError,
} from "@/generate/schema/http";

export function NewMediaRepository(golioApi: GolioApi): MediaRepository {
  return new media(golioApi);
}

class media implements MediaRepository {
  private golioApi: GolioApi;

  constructor(golioApi: GolioApi) {
    this.golioApi = golioApi;
  }
  async GetMedium(id: string): Promise<Medium> {
    let resp: MediaMediumIdGet200Response = {
      mediumId: "",
      thumbnailUrl: "",
      downloadUrl: "",
      fileType: "",
    };

    try {
      resp = await this.golioApi.mediaMediumIdGet({
        mediumId: id,
      });
    } catch (err) {
      if (err instanceof ResponseError) {
        if (err.response.status === 401 || err.response.status === 403) {
          throw new AuthError("failed get media", err);
        }
        throw new InternalError("failed get media", err);
      }
      throw err;
    }

    return {
      id: resp.mediumId,
      fileType: resp.fileType,
      downloadUrl: resp.downloadUrl,
    };
  }

  async FindMediumSummaries(
    offset: number,
    limit: number,
  ): Promise<FindMediumSummariesOutput> {
    let resp: MediaGet200Response = {
      totalCount: 0,
      media: [],
    };
    try {
      resp = await this.golioApi.mediaGet({
        offset: offset,
        limit: limit,
      });
    } catch (err) {
      if (err instanceof ResponseError) {
        if (err.response.status === 401 || err.response.status === 403) {
          throw new AuthError("failed find media", err);
        }
        throw new InternalError("failed find media", err);
      }
      throw err;
    }
    let mediumSummaries: MediumSummary[] = [];
    for (let medium of resp.media) {
      mediumSummaries.push({
        id: medium.id,
        fileType: medium.fileType,
        thumbnailUrl: medium.thumbnailUrl,
        createdAt: medium.createdAt,
        updatedAt: medium.updatedAt,
      });
    }
    return {
      totalCount: resp.totalCount,
      summaries: mediumSummaries,
    };
  }

  async CreateMedium(fileName: string): Promise<string> {
    let resp: MediaPost200Response = {
      uploadPresignedUrl: "",
    };
    try {
      resp = await this.golioApi.mediaPost({
        mediaPostRequest: {
          fileName: fileName,
        },
      });
    } catch (err) {
      if (err instanceof ResponseError) {
        if (err.response.status === 401 || err.response.status === 403) {
          throw new AuthError("failed create medium", err);
        }
        throw new InternalError("failed create medium", err);
      }
    }
    return resp.uploadPresignedUrl;
  }

  UploadFile(url: string, file: File): void {
    const xhr = new XMLHttpRequest();
    xhr.open("PUT", url, true);
    xhr.setRequestHeader("Content-Type", file.type);

    xhr.upload.onprogress = (event) => {
      if (event.lengthComputable) {
        const percentCompleted = Math.round((event.loaded * 100) / event.total);
        console.log("progress: ", percentCompleted);
      }
    };

    xhr.onload = () => {
      if (xhr.status === 200) {
        console.info("success file upload", file.name);
        return;
      }
      throw new Error(`failed upload file: ${file.name}`);
    };
    xhr.send(file);
  }
}
